### PR TITLE
Move `aws_smithy_http::operation::error::BuildError` into `aws-smithy-types`

### PR DESCRIPTION
**Pull request by ysaito1001**
## Motivation and Context
Takes care of the first part of https://github.com/awslabs/smithy-rs/issues/3054 (the remaining part is denoted as `TODO(runtimeCratesVersioningCleanup)` and until that's done the issue will not be closed).

## Description
This PR moves from `aws-smithy-http::operation::error::{BuildError, SerializationError}` to `aws-smithy-types::error::operation::{BuildError, SerializationError}`.

Within the origin crate, we leave "breadcrumbs" (i.e. reexports) for existing customers of the items above so they are not broken by the move. The breadcrumps will be removed by the part two of the issue.

As part of the move, `aws_smithy_http::endpoint::EndpointPrefix::new` now returns its `crate::endpoint::error::InvalidEndpointError` instead of `crate::operation::error::BuildError` for two reasons:
- `BuildError` is now in a stable crate and for an `InvalidUri` variant to be created, it needed to take a `http::uri::InvalidUri` from a unstable crate, which we cannot expose in public API.
- The new error is more closely related to the endpoint business.

## Testing
Relied on the tests in CI.

## Checklist
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

**Review by rcoh — COMMENTED** [review]
@jdisanti do you think this makes more sense in `aws-smithy-runtime-api`? I guess we need to define a litmus test of what goes where.

**Comment by ysaito1001** [comment]
> do you think this makes more sense in aws-smithy-runtime-api

Yeah, that distinction is blurry to me. For things like `BuildError` in particular, I don't have a guiding principle of which crate is a better home.

File: rust-runtime/aws-smithy-types/src/operation.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Types for representing the interaction between a service an a client, referred to as an "operation" in smithy.
+//! Clients "send" operations to services, which are composed of 1 or more HTTP requests.
+
+/// Errors for operations
+pub mod error {
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/operation.rs:10]: This should go under the existing root-level `error` module: `crate::error::operation`

Comment (ysaito1001) [reply rust-runtime/aws-smithy-types/src/operation.rs]: That arrangement is better at the new home. I seem to have focused too much on preserving the original module path. Addressed in f6bb4ba.

File: rust-runtime/aws-smithy-http/src/endpoint/error.rs [hunk 2]
```diff
@@ -69,23 +70,28 @@ pub struct InvalidEndpointError {
...
+    pub fn endpoint_must_have_scheme() -> Self {
         Self {
             kind: InvalidEndpointErrorKind::EndpointMustHaveScheme,
         }
     }
 
-    pub(super) fn failed_to_construct_authority(
+    /// Construct a build error for an invalid authority
+    pub fn failed_to_construct_authority(
+        authority: String,
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-http/src/endpoint/error.rs:82]: `impl Into<String>`?

**Review by jdisanti — APPROVED** [review]

File: rust-runtime/aws-smithy-http/src/endpoint/error.rs (hunk 2 above)
Comment (ysaito1001) [reply rust-runtime/aws-smithy-http/src/endpoint/error.rs]: Updated in 18e99d6.

**Pull request merged**